### PR TITLE
Tweak the way we handle the query…

### PR DESCRIPTION
**Pull request by Camwyn**
… so that event categories are in the expected format.

Code working, needs tests. Todos added.

:camera: http://p.tri.be/NoJWZA
:camera: http://p.tri.be/9iNnli
:camera: http://p.tri.be/vYj1Sc

[ECP-492]

[ECP-492]: https://moderntribe.atlassian.net/browse/ECP-492

### CODE REVIEW

File: src/Tribe/Views/V2/Hooks.php [hunk 1]
```diff
@@ -117,6 +119,29 @@ protected function add_filters() {
 		add_filter( 'tribe_events_views_v2_view_data', [ View_Utils::class, 'clean_data' ] );
 	}
 
+	/**
+	 * @todo: add docblock and revise this method.
+	 *
+	 * @since TBD
+	 *
+	 * @param WP $req
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Hooks.php:127]: :no_entry_sign: Missing parameter comment

<sub>hash: 286c78844738856a2ee3423a3c4d8823</sub>

File: src/Tribe/Views/V2/Hooks.php [hunk 2]
```diff
@@ -117,6 +119,29 @@ protected function add_filters() {
...
+	 * @return void
+	 */
+	public function parse_request( $req ) {
+
+		if ( empty( $req->query_vars[ TEC::TAXONOMY ] )  ) {
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Hooks.php:132]: :no_entry_sign: Expected exactly one space before closing parenthesis; "  " found.

<sub>hash: fe0236fcbb1e8d9e20636f3f1d94847d</sub>

File: src/Tribe/Views/V2/Hooks.php [hunk 3]
```diff
@@ -117,6 +119,29 @@ protected function add_filters() {
...
+			return;
+		}
+
+		if ( ! is_array( $req->query_vars[ TEC::TAXONOMY ] )  ) {
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Hooks.php:136]: :no_entry_sign: Expected exactly one space before closing parenthesis; "  " found.

<sub>hash: 08f791e7d5e938d19a793bfac55231cc</sub>

File: src/Tribe/Views/V2/View.php [hunk 4]
```diff
@@ -1146,7 +1148,20 @@ protected function setup_repository_args( Context $context = null ) {
 
 		// Set's up category URL for all views.
 		if ( ! empty( $context_arr[ TEC::TAXONOMY ] ) ) {
-			$args[ TEC::TAXONOMY ] = $context_arr[ TEC::TAXONOMY ];
+			if ( ! is_array( $context_arr[ TEC::TAXONOMY ] ) ) {
+				$args[ TEC::TAXONOMY ] = $context_arr[ TEC::TAXONOMY ];
+			} else {
+				if ( empty( $args['tax_query'] ) ) {
+					$args['tax_query'] = [];
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:1155]: :warning: Detected usage of tax_query, possible slow query.

<sub>hash: c2b29efd9bff4869a67fb6a0a436fe6f</sub>

File: src/Tribe/Views/V2/View.php [hunk 5]
```diff
@@ -1146,7 +1148,20 @@ protected function setup_repository_args( Context $context = null ) {
...
+				}
+
+				$args['tax_query'][] = [
+					'taxonomy' => TEC::TAXONOMY,
+					'field'    => 'slug',
+					'terms'    => $context_arr[ TEC::TAXONOMY ]
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:1161]: * :no_entry_sign: Each line in an array declaration must end in a comma
* :no_entry_sign: Each array item in a multi-line array declaration must end in a comma

<sub>hash: ec0cab995acc1dee11dba2cd150943dd</sub>

File: src/Tribe/Views/V2/View.php [hunk 6]
```diff
@@ -1146,7 +1148,20 @@ protected function setup_repository_args( Context $context = null ) {
...
+				];
+			}
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:1163]: * :no_entry_sign: Blank line found after control structure
* :no_entry_sign: Blank line found after control structure

<sub>hash: da8cd555fe1c75877aea59eaf11cf5ad</sub>

File: src/Tribe/Views/V2/View.php [hunk 7]
```diff
@@ -1849,19 +1864,27 @@ protected function get_breadcrumbs() {
...
 		$context_tax = $context->get( $taxonomy, false );
 
-		// Get term slug if taxonomy is not empty
+		// Get term slug if taxonomy is not empty.
 		if ( ! empty( $context_tax ) ) {
-			$term  = get_term_by( 'slug', $context_tax, $taxonomy );
-			if ( ! empty( $term->name ) ) {
-				$label = $term->name;
+			$context_tax = (array) $context_tax;
+			foreach( $context_tax as $term_slug ) {
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:1870]: * :no_entry_sign: Space after opening control structure is required
* :no_entry_sign: No space before opening parenthesis is prohibited
* :no_entry_sign: Expected 1 space(s) after FOREACH keyword; 0 found
* :no_entry_sign: There must be a space between foreach and the first parenthesis '('

<sub>hash: 8124d29e0533ca695fbf55c90873385e</sub>

**Review by tr1b0t — COMMENTED** [review]
:no_entry_sign: 11 error(s)
:warning: 1 warning(s)

**Comment by borkweb** [comment]
Should this be changed to a draft PR until the tests are in place so nobody spends time on the review too early in the process?

File: src/Tribe/Views/V2/Hooks.php [hunk 8]
```diff
@@ -117,6 +119,27 @@ protected function add_filters() {
...
+	 * @since TBD
+	 *
+	 * @param WP $req The WP Object we're trying to prepare for.
+	 */
+	public function parse_request( $req ) {
+		if ( empty( $req->query_vars[ TEC::TAXONOMY ] ) ) {
+			return;
+		}
+
+		if ( ! is_array( $req->query_vars[ TEC::TAXONOMY ] )  ) {
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Hooks.php:134]: :no_entry_sign: Expected exactly one space before closing parenthesis; "  " found.

<sub>hash: 451ddecc8cfd162e7f9c4a38ab27cd37</sub>

**Review by tr1b0t — COMMENTED** [review]
:no_entry_sign: 1 error(s)
:warning: 1 warning(s)

File: src/Tribe/Views/V2/View.php [hunk 9]
```diff
@@ -1146,7 +1146,19 @@ protected function setup_repository_args( Context $context = null ) {
 
 		// Set's up category URL for all views.
 		if ( ! empty( $context_arr[ TEC::TAXONOMY ] ) ) {
-			$args[ TEC::TAXONOMY ] = $context_arr[ TEC::TAXONOMY ];
+			if ( ! is_array( $context_arr[ TEC::TAXONOMY ] ) ) {
+				$args[ TEC::TAXONOMY ] = $context_arr[ TEC::TAXONOMY ];
+			} else {
+				if ( empty( $args['tax_query'] ) ) {
+					$args['tax_query'] = [];
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:1153]: :warning: Detected usage of tax_query, possible slow query.

<sub>hash: 314240cda0ff6d743adbbc155a93da12</sub>

**Review by tr1b0t — COMMENTED** [review]
:warning: 1 warning(s)

**Comment by Camwyn** [comment]
@lucatume @bordoni @borkweb Testing with taxonomies is not working in our test suite right now - so rather than hold this up for that, I can create a task ticket to add it and create tests for it separately?

**Comment by borkweb** [comment]
What do you mean by:the following?

> Testing with taxonomies is not working in our test suite right now

* Are you struggling with figuring out how to set up the tests that you need?
* Is something truly not functioning in the test suite?

Because of the areas that a change like this would touch, I would not consider this body of work complete nor ready for code review or QA until we have some test coverage to ensure we are not missing edge cases. If you need a buddy for this, I should be free after 3pm ET and can look at it with you.

**Comment by Camwyn** [comment]
@borkweb 
> * Are you struggling with figuring out how to set up the tests that you need?
> * Is something truly not functioning in the test suite?
> 

The functions we use to import test posts do not yet handle taxonomies - I can work around them (and am willing to do so) but it will take longer. I was offering an alternative.

I'll start on them and hit you up if (when 😉 ) I have issues?

File: src/Tribe/Views/V2/Hooks.php [hunk 10]
```diff
@@ -117,6 +118,27 @@ protected function add_filters() {
...
 
+	/**
+	 * Clear query vars that cause issues with the main query - they are handled elsewhere.
+	 *
+	 * @since TBD
+	 *
+	 * @param WP $req The WP Object we're trying to prepare for.
+	 */
+	public function parse_request( $req ) {
+		if ( empty( $req->query_vars[ TEC::TAXONOMY ] ) ) {
```
Comment (bordoni) [inline src/Tribe/Views/V2/Hooks.php:129]: When we were in our call I forgot, we should add some checks here to make sure it only happens on our Views V2 request. Like not on the admin or if anyone else is using.

Comment (Camwyn) [reply src/Tribe/Views/V2/Hooks.php:139]: ✅

File: tests/views_integration/Tribe/Events/Views/V2/Views/List_ViewTest.php [hunk 11]
```diff
@@ -101,6 +102,168 @@ public function test_render_with_upcoming_events() {
...
+
+		$list_view = View::make( List_View::class );
+		$context = tribe_context()->alter( [
+			'event_display'      => 'list',
+			'event_display_mode' => 'list',
+			'today'              => $this->mock_date_value,
+			'now'                => $this->mock_date_value,
+			'event_date'         => $this->mock_date_value,
+			'events_per_page'    => 3,
+			'tax_input'  => [
```
Comment (lucatume) [inline tests/views_integration/Tribe/Events/Views/V2/Views/List_ViewTest.php:170]: This is the part that puzzles me a bit and where I think this fix is not the best one.
In our Repository code we should support the `category` argument for the `where` or `by` method.
Example:

```php
$events = tribe_events()->where('category', 'foo')->get_ids();
```

And, if any fix is required to make that work, then it should be done in the Events repository to make sure it works.

Since the Context and the Repository are "neutral" to unknown inputs, as is the `tax_input` here, they will just pass them through to the query and, sure, the `tax_input` query argument will work.
This seems like a workaround for not working code (the Repository one handling the `category` criteria), and not a fix.

Furthermore, what if the `tax_input` already contains other criteria? Would they be kept or overridden?

**Review by lucatume — CHANGES_REQUESTED** [review]

**Comment by Camwyn** [comment]
Closing in favor of https://github.com/moderntribe/the-events-calendar/pull/3304
